async_simple/coro/Dispatch.h

```diff
@@ -0,0 +1,116 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef ASYNC_SIMPLE_CORO_DISPATCH_H
+#define ASYNC_SIMPLE_CORO_DISPATCH_H
+
+#include "async_simple/Common.h"
+#include "async_simple/Executor.h"
+#include "async_simple/coro/Lazy.h"
+#include "async_simple/experimental/coroutine.h"
+
+#include <cassert>
+#include <string>
+
+#include <iostream>
+
+namespace async_simple {
+namespace coro {
+
+namespace detail {
+
+// based on c++ coroutine common abi.
+// it is not supported that the PromiseType's align is larger than the size of
+// two function pointer.
+inline std::coroutine_handle<> GetContinuationFromHandle(
+    std::coroutine_handle<> h) {
+    constexpr size_t func_offset = 2 * sizeof(void (*)(void*));
+    char* ptr = static_cast<char*>(h.address());
+    ptr = ptr + func_offset;
+    return *reinterpret_cast<std::coroutine_handle<>*>(ptr);
+}
+
+inline void ChangeCoroutinesExecutorTo(std::coroutine_handle<> h,
+                                       Executor* ex) {
+    while (true) {
+        std::coroutine_handle<> continuation = GetContinuationFromHandle(h);
+        if (!continuation.address()) {
+            break;
+        } else {
+            auto& promise =
+                std::coroutine_handle<LazyPromiseBase>::from_address(
+                    h.address())
+                    .promise();
+            promise._executor = ex;
+        }
+        h = continuation;
+    }
+}
+
+class DispatchAwaiter {
+public:
+    explicit DispatchAwaiter(Executor* ex) noexcept : _ex(ex) {
+        assert(_ex != nullptr);
+    }
+
+    bool await_ready() const noexcept { return false; }
+
+    template <typename PromiseType>
+    bool await_suspend(std::coroutine_handle<PromiseType> h) {
+        if (h.promise()._executor == _ex) {
+            return false;
+        }
+        Executor* old_ex = h.promise()._executor;
+        ChangeCoroutinesExecutorTo(h, _ex);
+        std::string err_msg =
+            std::string("dispatch to executor ") + _ex->name() + " failed";
+        bool succ = _ex->schedule(std::move(h));
+        // cannot access *this after schedule
+        if (succ == false) {
+            ChangeCoroutinesExecutorTo(h, old_ex);
+        }
+        logicAssert(succ, err_msg.c_str());
+        return true;
+    }
+
+    void await_resume() noexcept {}
+
+private:
+    Executor* _ex;
+};
+
+class DispatchAwaitable {
+public:
+    explicit DispatchAwaitable(Executor* ex) : _ex(ex) {}
+
+    auto coAwait(Executor*) { return DispatchAwaiter(_ex); }
+
+private:
+    Executor* _ex;
+};
+
+}  // namespace detail
+
+// schedule this coroutine to ex for execution
```

Comment:
```suggestion
// schedule this Lazy to ex for execution
```

这个by design只能给 Lazy 使用